initialization/vrf_search.go

```diff
@@ -0,0 +1,183 @@
+	options := searchForNonceOpts{
+		logger:            zap.NewNop(),
+		powDifficultyFunc: shared.PowDifficulty,
+	}
+	for _, opt := range opts {
+		opt(&options)
+	}
+	logger := options.logger
+
+	metadata, err := LoadMetadata(initOpts.DataDir)
+	if err != nil {
+		return 0, nil, fmt.Errorf("failed to load metadata: %w", err)
+	}
+
+	difficulty := options.powDifficultyFunc(metadata.LabelsPerUnit * uint64(metadata.NumUnits))
+	logger.Info("searching for lowes nonce",
+		zap.String("difficulty", hex.EncodeToString(difficulty)),
+		zap.String("datadir", initOpts.DataDir),
+	)
+
+	allFiles, err := os.ReadDir(initOpts.DataDir)
+	if err != nil {
+		return 0, nil, fmt.Errorf("couldn't open the data directory: %w", err)
+	}
+
+	woReference, err := oracle.New(
+		oracle.WithProviderID(CPUProviderID()),
+		oracle.WithCommitment(metadata.CommitmentAtxId),
+		oracle.WithVRFDifficulty(difficulty),
+		oracle.WithScryptParams(initOpts.Scrypt),
+		oracle.WithLogger(logger),
+	)
+	if err != nil {
+		return 0, nil, err
+	}
+	defer woReference.Close()
+
+	for _, file := range allFiles {
+		info, err := file.Info()
+		if err != nil {
+			logger.Error("failed to get file info", zap.Error(err))
+			continue
+		}
+
+		if !shared.IsInitFile(info) {
+			continue
+		}
+
+		fileIndex, err := shared.ParseFileIndex(file.Name())
+		if err != nil {
+			logger.Panic("failed to parse file index", zap.Error(err))
+		}
+		firstLabelIndex := uint64(fileIndex) * metadata.MaxFileSize / postrs.LabelLength
+		logger.Info("looking for VRF nonce in file", zap.String("file", file.Name()))
+
+		file, err := os.Open(filepath.Join(initOpts.DataDir, file.Name()))
+		if err != nil {
+			logger.Error("failed to open file", zap.Error(err))
+			continue
+		}
+		defer file.Close()
+
+		idx, l, err := searchForNonce(ctx, bufio.NewReader(file), difficulty, woReference)
+		if err != nil {
+			return 0, nil, fmt.Errorf("failed to search for nonce: %w", err)
+		}
+		if l != nil {
+			nonceValue = l
+			nonce = firstLabelIndex + idx
+			logger.Info("found nonce", zap.Uint64("index", nonce), zap.String("label", hex.EncodeToString(nonceValue)))
+			difficulty = nonceValue // override difficulty to the new lowest label
+
+		}
+	}
+	if nonceValue != nil {
+		logger.Info("updating postdata_metadata.json with found nonce", zap.Uint64("nonce", nonce), zap.String("NonceValue", hex.EncodeToString(nonceValue)))
+		metadata.Nonce = &nonce
+		metadata.NonceValue = shared.NonceValue(nonceValue)
+		SaveMetadata(initOpts.DataDir, metadata)
+		return nonce, nonceValue, nil
+	}
+	return 0, nil, ErrNonceNotFound
+}
+
+// searchForNonce searches for a nonce in the given reader.
+func searchForNonce(ctx context.Context, r io.Reader, difficulty []byte, oracle *oracle.WorkOracle) (nonce uint64, nonceValue []byte, err error) {
+	labelBuf := make([]byte, postrs.LabelLength)
+	labelIndex := uint64(0)
+	for {
+		select {
+		case <-ctx.Done():
+			return nonce, labelBuf, ctx.Err()
```

Comment:
The way it is implemented at the moment if once ctrl+C the search after a nonce was found it won't be saved. Only after the search is completed.

Intended?

Reply:
Thanks, good catch :+1: